Additions and changes for trailers

neithernut opened the pull request.

matthiasbeyer committed:
Add trailer module

matthiasbeyer committed:
Add IntoMap trait and implement on iterators

matthiasbeyer committed:
fixup! Add IntoMap trait and implement on iterators

matthiasbeyer committed:
Add Into tuple for Trailer

neithernut committed:
Add module for message related functionality

neithernut committed:
Move trailers module into message module

Trailers are message parts.

neithernut committed:
Implement From<String> for TrailerKey

This will let us construct `TrailerKey`s from other modules.

neithernut committed:
Make TralerValue::from_slice public

This will enable construction of `TralerValue`s from other modules.

neithernut committed:
Add initial version of "lines" module

This module will enable parsing single lines from a message.

neithernut committed:
Rewrite Trailers using the new Line type

This also makes Trailers skip blocks of non-blank lines containing
non-trailers (e.g. regular lines of text).

neithernut committed:
Add function for appending to a TrailerValue

This function will be useful for assembling multi line trailers.

neithernut committed:
Add Lines iterator for iterating over different lines

neithernut commented:
The implementation of `Trailers::next()` could be improved. Documentation is missing, too.

matthiasbeyer commented on lib/src/message/line.rs:
```diff
@@ -0,0 +1,79 @@
+use message::trailer::{Trailer, TrailerKey, TrailerValue};
+use regex::Regex;
+use std::iter::Peekable;
+use std::str::Lines as L;
```
Don't rename this, it is absolutely confusing reading `L` in the code and not finding an generic around. You use it only one time anyways.

matthiasbeyer commented on lib/src/message/line.rs:
```diff
@@ -0,0 +1,79 @@
+    fn from(line :&'a str) -> Self {
+        lazy_static! {
+            // TODO: do not accept whitespace in the key
+            static ref RE: Regex = Regex::new(r"^(?P<key>(.*)):\ (?P<value>(.*))$").unwrap();
```
I guess `[[:^space:]]` [would do](https://docs.rs/regex/0.2.1/regex/).

matthiasbeyer commented on lib/src/message/line.rs:
```diff
@@ -0,0 +1,79 @@
+        }
+
+        if let Some(cap) = RE.captures(trimmed) {
+            if let (Some(key), Some(value)) = (cap.name("key"), cap.name("value")) {
```
Could it be that

```rust
if let (Some(Key), Some(value)) = RE.captures(trimmed).map(|c| (c.name("key"), c.name("value))) {
```

would do?

matthiasbeyer commented on lib/src/message/line.rs:
```diff
@@ -0,0 +1,79 @@
+
+
+#[derive(Debug, Clone)]
+pub struct Lines<'a>(Peekable<L<'a>>);
```
(See above.)

matthiasbeyer commented on lib/src/message/line.rs:
```diff
@@ -0,0 +1,79 @@
+}
+
+
+#[derive(Debug, Clone)]
```
I am not sure whether we need `Clone` here. Also I'm not sure whether this would actually work as intended...

matthiasbeyer commented on lib/src/message/line.rs:
```diff
@@ -0,0 +1,79 @@
+    fn next(&mut self) -> Option<Self::Item> {
+        match self.0.next().map(|line| Line::from(line)) {
+            Some(Line::Trailer(trailer)) => {
+                let mut t = trailer;
```
I guess you can leave this out if you match against `Some(Line::Trailer(mut trailer))`.

matthiasbeyer commented on lib/src/message/trailer.rs:
```diff
@@ -32,6 +39,12 @@ impl TrailerValue {
         }
     }
 
+    pub fn append(self, slice: &str) -> TrailerValue {
```
Maybe a comment that this automatically turns everything into a `TrailerValue::String`?

matthiasbeyer commented on lib/src/message/trailer.rs:
```diff
@@ -65,31 +84,24 @@ impl<'a> Iterator for Trailers<'a> {
+            }
+
+            // refill buffer from next block
+            loop {
```
I would use [loop labels](https://doc.rust-lang.org/book/loops.html#loop-labels) so one can easily see which loop a `break` is for.

neithernut replied:
Still not sure whether I'll rewrite the entire method.

neithernut replied:
The thing has to be rewritten because the logic for text lines is erroneous.

matthiasbeyer reviewed (changes requested):
Besides the changes you suggested, here are my annotations.

neithernut committed:
Implement From<std::Lines> for message::Lines

This will come in handy in situations where we want to extract `Line`
structs for only a part of a message.

neithernut committed:
Use From<std::Lines> in message::Lines::new()

neithernut commented:
I'm thinking of introducing yet another type in order to simplify `Trailers`: a `TrailerCollector` enum to encapsulate the state necessary for scanning a block. A value `Collecting` would wrap a `LinkedList` (maybe as a reference) for collection. A function would append to the wrapped list or ignore the argument for other cases.

matthiasbeyer commented on lib/src/message/line.rs:
```diff
@@ -22,35 +28,37 @@ pub enum Line {
         lazy_static! {
-            // TODO: do not accept whitespace in the key
-            static ref RE: Regex = Regex::new(r"^(?P<key>(.*)):\ (?P<value>(.*))$").unwrap();
+            static ref RE: Regex = Regex::new(r"^(?P<key>([^[:space]]+)):\ (?P<value>(.*))$").unwrap();
```
Missing ending `:` in char class `space`.

neithernut committed:
Add enum type for colecting trailer values

This will simplify the implementation of `Trailers::next()`.

matthiasbeyer commented:
Ready to squash from my POV.

neithernut commented:
Done.

matthiasbeyer commented:
Was ff-merged in https://github.com/neithernut/git-dit/pull/35 - doesn't close here because of rebase.
